ext/storage_sources/s3_store/s3_store.cxx

```diff
@@ -59,6 +65,151 @@ static int s3_customize_file_system(
   WT_STORAGE_SOURCE *, WT_SESSION *, const char *, const char *, const char *, WT_FILE_SYSTEM **);
 static int s3_add_reference(WT_STORAGE_SOURCE *);
 static int s3_fs_terminate(WT_FILE_SYSTEM *, WT_SESSION *);
+static int s3_get_directory(const char *home, const char *s, ssize_t len, bool create, char **copy);
+static int s3_stat(WT_FILE_SYSTEM *file_system, WT_SESSION *session, const char *name,
+  const char *caller, bool must_exist, struct stat *statp);
+static int s3_cache_path(WT_FILE_SYSTEM *file_system, const char *name, char **pathp);
+static int s3_path(WT_FILE_SYSTEM *file_system, const char *dir, const char *name, char **pathp);
+static int s3_exist(
+  WT_FILE_SYSTEM *file_system, WT_SESSION *session, const char *name, bool *existp);
+
+/*
+ * s3_exist --
+ *     Return if the file exists.
+ */
+static int
+s3_exist(WT_FILE_SYSTEM *file_system, WT_SESSION *session, const char *name, bool *existp)
+{
+    struct stat sb;
+    S3_STORAGE *s3;
+    int ret;
+
+    s3 = FS2S3(file_system);
+    s3->op_count++;
+    *existp = false;
+
+    if ((ret = s3_stat(file_system, session, name, "ss_exist", false, &sb)) == 0)
+        *existp = true;
+    else if (ret == ENOENT) {
+        ret = 0;
+    }
+    return (ret);
+}
+
+/*
+ * s3_path --
+ *     Construct a pathname from the file system and the object name.
+ */
+static int
+s3_path(WT_FILE_SYSTEM *file_system, const char *dir, const char *name, char **pathp)
+{
+    size_t len;
+    int ret;
+    char *p;
+    ret = 0;
+
+    /* Skip over "./" and variations (".//", ".///./././//") at the beginning of the name. */
+    while (*name == '.') {
+        if (name[1] != '/')
+            break;
+        name += 2;
+        while (*name == '/')
+            name++;
+    }
+    len = strlen(dir) + strlen(name) + 2;
+    if ((p = (char *)malloc(len)) == NULL)
+        /* Out of memory */
+        return (ENOMEM);
+    if (snprintf(p, len, "%s/%s", dir, name) >= (int)len)
+        /* Overflow sprintf */
+        return (EINVAL);
+    *pathp = p;
+    return (ret);
+}
+
+/*
+ * s3_cache_path --
+ *     Construct the path to the object in the cache from the file system and the object name.
+ */
+static int
+s3_cache_path(WT_FILE_SYSTEM *file_system, const char *name, char **pathp)
+{
+    return (s3_path(file_system, ((S3_FILE_SYSTEM *)file_system)->cache_dir, name, pathp));
+}
+
+/*
+ * s3_stat --
+ *     Perform the stat system call for a name in the file system. If the file is not present in the
+ *     local cache, the S3 Bucket is checked.
+ */
+static int
+s3_stat(WT_FILE_SYSTEM *file_system, WT_SESSION *session, const char *name, const char *caller,
```

Comment:
What is caller for?

Reply:
I think it's for error handling and logging. Looking at the code in local_store; it looks like it just used for debugging - https://github.com/wiredtiger/wiredtiger/blob/71a41e1a2b84e7bd8af1f4e8c8a381ad1b882e83/ext/storage_sources/local_store/local_store.c#L404

Reply:
https://github.com/wiredtiger/wiredtiger/blob/71a41e1a2b84e7bd8af1f4e8c8a381ad1b882e83/ext/storage_sources/local_store/local_store.c#L433